activation/nipost.go

```diff
@@ -440,13 +445,14 @@ func (nb *NIPostBuilder) submitPoetChallenges(
 			)
 		case len(existingRegistrations) == 0:
 			// no existing registration for given poets set
-			nb.logger.Panic(
+			nb.logger.Warn(
 				"None of the poets listed in the config matches the existing registrations. "+
 					"Verify your config and local database state.",
 				zap.Strings("registrations", maps.Keys(registrationsMap)),
 				zap.Strings("configured_poets", maps.Keys(nb.poetProvers)),
 				log.ZShortStringer("smesherID", nodeID),
 			)
+			return nil, ErrNoRegistrationForGivenPoetFound
```

Comment:
Instead of

```go
logger.Fatal("submitting to poets", zap.Error(err))
```

This log line should be at the caller site otherwise this error will be logged twice. In general we should always either return/handle an error or log it and avoid doing both 🙂

Reply:
May be I understood you wrong, but if you want to put log.Fatal() here, check the discussion
https://github.com/spacemeshos/go-spacemesh/pull/6116#discussion_r1701432124

And actually error is not logged twice: I log the info about invalid state + log error later.

Reply:
What I mean is we should avoid logging and returning an error in general. It leads to very verbose and repeating log messages. If `len(existingRegistrations) == 0` we log here a warning and return this error, the only caller of this function then checks for this specific error and logs the same information again on fatal level - so you get:

```
WARN  none of poets listed ...
FATAL submitting to poets ... error: "builder: none of configured poets matches the existing registrations"
```

While I think it should instead be

```
FATAL  none of poets listed ...
```

So instead I suggest to only return the error here and on the caller site change the `"submitting to poets"` log message to the log message here and not logging the error again (it is already handled by logging the "None of the poets listed..." message)

Reply:
Yes, but:
- we were talking about proper user notification. I think, it is important to show, which poets were used for registration and which poets user has now. So to log it with Fatal() on caller side, we have to return registrations from db + error, what I think is strange. Because we return value only to log it. 
- wrapping error with registrations info and so on also is not correct, because error should not include the whole information like log message (here I am agree with Elad)

Reply:
> wrapping error with registrations info and so on also is not correct, because error should not include the whole information like log message (here I am agree with Elad)

Yes, it makes no sense to for the error to be more than just information about what went wrong, but this can include the context needed as to how it went wrong:

Instead of `ErrNoRegistrationForGivenPoetFound` you could have a custom type `PoetRegistrationMismatchError`:

```go
type PoetRegistrationMismatchError struct {
	Registrations []string
	ConfiguredPoets []string
}

func (e PoetRegistrationMismatchError) Error() string {
	var sb strings.Builder
	sb.WriteString("builder: none of configured poets matches the existing registrations.\n")
	sb.WriteString("registrations:\n")
	for _, r := range e.Registrations {
		sb.WriteString("\t")
		sb.WriteString(r)
		sb.WriteString("\n")
	}
	sb.WriteString("\nconfigured poets:\n")
	for _, p := range e.ConfiguredPoets {
		sb.WriteString("\t")
		sb.WriteString(p)
		sb.WriteString("\n")
	}
	return sb.String()
}
```

returning it here like this:

```go
		case len(existingRegistrations) == 0:
			// no existing registration for given poets set
			return nil, &PoetRegistrationMismatchError{
				Registrations:   maps.Keys(registrationsMap),
				ConfiguredPoets: maps.Keys(nb.poetProvers),
			}
```

and on the caller side you can change the error handling to:

```go
	submittedRegistrations, err := nb.submitPoetChallenges(
		ctx,
		signer,
		poetProofDeadline,
		poetRoundStart, challenge.Bytes(),
	)
	regErr := &PoetRegistrationMismatchError{}
	switch {
	case errors.As(err, &regErr):
		logger.Fatal(
			"None of the poets listed in the config matches the existing registrations. "+
				"Verify your config and local database state.",
			zap.Strings("registrations", regErr.Registrations),
			zap.Strings("configured_poets", regErr.ConfiguredPoets),
		)
	default:
		return nil, fmt.Errorf("submitting to poets: %w", err)
	}
```